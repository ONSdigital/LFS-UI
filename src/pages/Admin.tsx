import React, {Component, Fragment} from "react";
import {roleHeaders, userHeaders} from "../utilities/Headers";
import {ONSAccordionTable} from "../components/ONS_DesignSystem/ONSAccordionTable";
import {ONSButton} from "../components/ONS_DesignSystem/ONSButton";
import ReactModal from "react-modal";
import {ONSTextInput} from "../components/ONS_DesignSystem/ONSTextInput";
import DocumentTitle from "react-document-title";

interface State {
    UserData: Data | null
    Users: any[] | null
    RoleData: any[] | null
    showSaveModal: boolean
}

interface Data {
    Rows: Row,
    Count: number
}

interface Row {
    [key: number]: Cell
}

interface Cell {
    [key: string]: object
}

export class Admin extends Component <{}, State> {
    displayName = Admin.name;

    constructor(props: any) {
        super(props);
        this.state = {UserData: null, Users: [], RoleData: [], showSaveModal: false};
        this.getRoles();
        this.getUsers();
    }

    //user table is created in TableWithModal so all stuff is there
    getRoles = () => {
        fetch("/jsons/Roles.json")
            .then(response => response.json())
            .then(response => this.setMockRoleData(response));
    };

    setMockRoleData = (response: any) => {
        this.setState({RoleData: response.Rows});
    };

    getUsers = () => {
        fetch("/jsons/Users.json")
            .then(response => response.json())
            .then(response => this.setMockUserData(response));
    };

    setMockUserData = (response: any) => {
        this.setState({Users: response.Rows});
    };

    createUser = (payload: any) => {
        console.log(payload);
    };

    //save modal functions
    openSaveModal = () => {
        this.setState({showSaveModal: true});
    };

    closeSaveModal = () => {
        this.setState({showSaveModal: false});
    };

    saveModal = () => {
        if (this.state.showSaveModal) {
            return (
                // Modal was messinhg with the submit button so we made our own and its wayyy cooler
                <ReactModal
                    key={1}
                    isOpen={this.state.showSaveModal}
                    contentLabel="Minimal Modal Example"
                    className='Modal'
                    shouldFocusAfterRender={true}
                    shouldReturnFocusAfterClose={true}
                >
                    <h1>Add User</h1>
                    {userHeaders().map((header, index) => {
                            return header.create &&
                                (
                                    <Fragment key={index}>
                                        <ONSTextInput label={header.label}/>
                                    </Fragment>
                                );
                        }
                    )}
                    <br/>
                    <ONSButton primary={true} small={false} label={" Save "} onClick={this.createUser}/>
                    <ONSButton label="Cancel" small={false} primary={false} onClick={this.closeSaveModal}/>
                </ReactModal>
            );
        }
    };

    RoleTableRow = (rowData: any) => {
        let row = rowData.row;
        return (
            <>
                <td className="table__cell ">
                    {row.Name}
                </td>
                <td className="table__cell ">
                    {row.Pages.map(((item: string) => item + " , "))}
                </td>
            </>
        );
    };

    UsersTableRow = (rowData: any) => {
        let row = rowData.row;
        return (
            <>
                <td className="table__cell ">
                    {row.user_id}
                </td>
                <td className="table__cell ">
                    {row.username}
                </td>
                <td className="table__cell ">
                    {row.password}
                </td>
                <td className="table__cell ">
                    {row.role}
                </td>
            </>
        );
    };

    render() {
        return (
            <DocumentTitle title={"LFS User Management"}>
                <div className="container">
                    <br></br>
                    {[this.saveModal()]}
                    <ONSAccordionTable Headers={roleHeaders()}
                                       data={this.state.RoleData}
                                       Row={this.RoleTableRow}
                                       expandedRowEnabled={false}
                                       noDataMessage={"No Roles Found"}/>
                    <ONSButton label={"New User"} primary={false} onClick={this.openSaveModal}/>
                    <hr/>
                    <ONSAccordionTable Headers={userHeaders()}
                                       data={this.state.Users}
                                       Row={this.UsersTableRow}
                                       expandedRowEnabled={false}
                                       pagination={true}
                                       paginationSize={10}
                                       noDataMessage={"No Users Found"}/>
                </div>
            </DocumentTitle>
        );
    }
}
