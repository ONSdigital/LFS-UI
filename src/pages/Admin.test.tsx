import React from "react";
import {cleanup, fireEvent, render} from "@testing-library/react";

import {BrowserRouter} from "react-router-dom";

import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import fetch from "../tests/setup/__mocks__/fetch.js";
import {Admin} from "./Admin";
import MockDate from "mockdate";

// @ts-ignore
global.fetch = fetch;
MockDate.set('1955-06-18');


describe("Admin - User Management Page", () => {
    Enzyme.configure({adapter: new Adapter()});

    beforeEach(() => {
        // @ts-ignore
        process.env.NODE_ENV = "test";
    });

    afterEach(() => {
        cleanup();
    });

    function wrapper(render: any) {
        return render(
            <BrowserRouter>
                <Admin/>
            </BrowserRouter>
        );
    }

    it("renders the page", async () => {
        expect(wrapper(render)).toMatchSnapshot();
    });

    it("should open and close the New User Modal ", async () => {
        const {getByLabelText, getByText} = wrapper(render);

        await fireEvent.click(getByText(/New User/i));

        await fireEvent.click(getByText(/Cancel/i));

        await fireEvent.click(getByText(/New User/i));

        await fireEvent.change(getByLabelText(/Username/i), {
            target: {value: "Admin"}
        });

        await fireEvent.click(getByText(/Save/i));
    });
});