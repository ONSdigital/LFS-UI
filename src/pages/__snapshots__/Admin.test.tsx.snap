// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin - User Management Page renders the page 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <br />
        <table
          class="table table--sortable"
          data-aria-asc="ascending"
          data-aria-desc="descending"
          data-aria-sort="Sort by"
          id="basic-table"
        >
          <thead
            class="table__head"
          >
            <tr
              class="table__row"
            >
              <th
                aria-sort="none"
                class="table__header "
                scope="col"
              >
                <button
                  aria-label="Sort by Role Name"
                  class="table__sort-button"
                  data-index="0"
                  type="button"
                >
                  Role Name
                </button>
              </th>
              <th
                class="table__header "
                scope="col"
              >
                Page Access
              </th>
            </tr>
          </thead>
          <tbody
            class="table__body "
          >
            <tr>
              <td
                class="table__cell "
                colspan="2"
              >
                <div
                  class="panel panel--info panel--simple "
                >
                  <div
                    class="panel__body"
                  >
                    <p>
                      No Roles Found
                    </p>
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
        <button
          class="btn btn--secondary "
          data-testid="button"
          type="button"
        >
          <span
            class="btn__inner"
          >
            New User
          </span>
        </button>
        <hr />
        <table
          class="table table--sortable"
          data-aria-asc="ascending"
          data-aria-desc="descending"
          data-aria-sort="Sort by"
          id="basic-table"
        >
          <thead
            class="table__head"
          >
            <tr
              class="table__row"
            >
              <th
                aria-sort="none"
                class="table__header "
                scope="col"
              >
                <button
                  aria-label="Sort by User_ID"
                  class="table__sort-button"
                  data-index="0"
                  type="button"
                >
                  User_ID
                </button>
              </th>
              <th
                aria-sort="none"
                class="table__header "
                scope="col"
              >
                <button
                  aria-label="Sort by Username"
                  class="table__sort-button"
                  data-index="1"
                  type="button"
                >
                  Username
                </button>
              </th>
              <th
                class="table__header "
                scope="col"
              >
                Password
              </th>
              <th
                aria-sort="none"
                class="table__header "
                scope="col"
              >
                <button
                  aria-label="Sort by Role"
                  class="table__sort-button"
                  data-index="3"
                  type="button"
                >
                  Role
                </button>
              </th>
            </tr>
          </thead>
          <tbody
            class="table__body "
          >
            <tr>
              <td
                class="table__cell "
                colspan="4"
              >
                <div
                  class="panel panel--info panel--simple "
                >
                  <div
                    class="panel__body"
                  >
                    <p>
                      No Users Found
                    </p>
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
        <div
          hidden=""
        >
          <nav
            aria-label="Pagination (Page 1 of 0)"
            class="pagination "
          >
            <div
              class="pagination__position"
            >
              Page 
              1
               of 
              0
            </div>
            <ul
              class="pagination__items"
              style="margin-left: 0px;"
            >
              <li
                class="pagination__item pagination__item--current"
                style="cursor: pointer;"
              >
                <button
                  aria-label="Go to page 1"
                  class="pagination__link"
                  id="link"
                  style="cursor: pointer; outline: 0; text-decoration: underline;"
                >
                  1
                </button>
              </li>
              <li
                class="pagination__item pageination__item--gap"
              >
                …
              </li>
              <li
                class="pagination__item pageination__item--gap"
              >
                …
              </li>
              <li
                class="pagination__item "
                style="cursor: pointer;"
              >
                <button
                  aria-label="Go to page 0"
                  class="pagination__link"
                  id="link"
                  style="cursor: pointer; outline: 0; text-decoration: underline;"
                >
                  0
                </button>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <br />
      <table
        class="table table--sortable"
        data-aria-asc="ascending"
        data-aria-desc="descending"
        data-aria-sort="Sort by"
        id="basic-table"
      >
        <thead
          class="table__head"
        >
          <tr
            class="table__row"
          >
            <th
              aria-sort="none"
              class="table__header "
              scope="col"
            >
              <button
                aria-label="Sort by Role Name"
                class="table__sort-button"
                data-index="0"
                type="button"
              >
                Role Name
              </button>
            </th>
            <th
              class="table__header "
              scope="col"
            >
              Page Access
            </th>
          </tr>
        </thead>
        <tbody
          class="table__body "
        >
          <tr>
            <td
              class="table__cell "
              colspan="2"
            >
              <div
                class="panel panel--info panel--simple "
              >
                <div
                  class="panel__body"
                >
                  <p>
                    No Roles Found
                  </p>
                </div>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
      <button
        class="btn btn--secondary "
        data-testid="button"
        type="button"
      >
        <span
          class="btn__inner"
        >
          New User
        </span>
      </button>
      <hr />
      <table
        class="table table--sortable"
        data-aria-asc="ascending"
        data-aria-desc="descending"
        data-aria-sort="Sort by"
        id="basic-table"
      >
        <thead
          class="table__head"
        >
          <tr
            class="table__row"
          >
            <th
              aria-sort="none"
              class="table__header "
              scope="col"
            >
              <button
                aria-label="Sort by User_ID"
                class="table__sort-button"
                data-index="0"
                type="button"
              >
                User_ID
              </button>
            </th>
            <th
              aria-sort="none"
              class="table__header "
              scope="col"
            >
              <button
                aria-label="Sort by Username"
                class="table__sort-button"
                data-index="1"
                type="button"
              >
                Username
              </button>
            </th>
            <th
              class="table__header "
              scope="col"
            >
              Password
            </th>
            <th
              aria-sort="none"
              class="table__header "
              scope="col"
            >
              <button
                aria-label="Sort by Role"
                class="table__sort-button"
                data-index="3"
                type="button"
              >
                Role
              </button>
            </th>
          </tr>
        </thead>
        <tbody
          class="table__body "
        >
          <tr>
            <td
              class="table__cell "
              colspan="4"
            >
              <div
                class="panel panel--info panel--simple "
              >
                <div
                  class="panel__body"
                >
                  <p>
                    No Users Found
                  </p>
                </div>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
      <div
        hidden=""
      >
        <nav
          aria-label="Pagination (Page 1 of 0)"
          class="pagination "
        >
          <div
            class="pagination__position"
          >
            Page 
            1
             of 
            0
          </div>
          <ul
            class="pagination__items"
            style="margin-left: 0px;"
          >
            <li
              class="pagination__item pagination__item--current"
              style="cursor: pointer;"
            >
              <button
                aria-label="Go to page 1"
                class="pagination__link"
                id="link"
                style="cursor: pointer; outline: 0; text-decoration: underline;"
              >
                1
              </button>
            </li>
            <li
              class="pagination__item pageination__item--gap"
            >
              …
            </li>
            <li
              class="pagination__item pageination__item--gap"
            >
              …
            </li>
            <li
              class="pagination__item "
              style="cursor: pointer;"
            >
              <button
                aria-label="Go to page 0"
                class="pagination__link"
                id="link"
                style="cursor: pointer; outline: 0; text-decoration: underline;"
              >
                0
              </button>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
